lib/Conversion/TorchToLinalg/Pooling.cpp

```diff
@@ -829,226 +830,333 @@ class ConvertAtenMaxUnpool3dOp final
-      return rewriter.notifyMatchFailure(
-          op, "unimplemented: count_include_pad is expected to be true for "
-              "AtenAvgPool3dOp");
-    }
+  constexpr int avgPoolDims = getAvgPoolNumOfDims<OpTy>();
 
-    // Case2: AtenAvgPool1/3dOp without count_include_pad equal to `False`.
-    divisor = kernelSizeIntValues[0];
-    for (uint32_t i = 1; i < kernelSizeIntValues.size(); i++) {
-      divisor =
-          rewriter.create<arith::MulIOp>(loc, divisor, kernelSizeIntValues[i]);
-    }
-    if constexpr (!std::is_same<OpTy, AtenAvgPool1dOp>()) {
-      divisor = isa<Torch::NoneType>(op.getDivisorOverride().getType())
-                    ? divisor
-                    : adaptor.getDivisorOverride();
-    }
-    divisor = convertScalarToDtype(rewriter, loc, divisor, resultElementType);
-    avgPool = rewriter
-                  .create<linalg::GenericOp>(
-                      loc, outputTensor.getType(), sumPool, outputTensor,
-                      /*indexingMaps=*/indexingMapsAvg,
-                      /*iteratorTypes=*/iteratorTypesAvg,
-                      [&](OpBuilder &b, Location loc, ValueRange args) {
-                        Value avg;
-                        if (isa<mlir::IntegerType>(resultElementType))
-                          avg = b.create<arith::DivSIOp>(loc, args[0], divisor);
-                        else if (isa<mlir::FloatType>(resultElementType))
-                          avg = b.create<arith::DivFOp>(loc, args[0], divisor);
-                        b.create<linalg::YieldOp>(loc, avg);
-                      })
-                  .getResult(0);
-    rewriter.replaceOpWithNewOp<tensor::CastOp>(op, resultType, avgPool);
-    return success();
+  bool noPadding = llvm::all_of(paddingInts, [](int64_t p) { return p == 0; });
+  if (countIncludePad || noPadding) {
+    // These cases are not handled here.
+    return std::nullopt;
   }
-};
-} // namespace
+  if (avgPoolDims < 1) {
+    return rewriter.notifyMatchFailure(
+        op, "Unexpected type. Only expected AtenAvgPool1dOp, AtenAvgPool2dOp, "
+            "and AtenAvgPool3dOp.");
+  }
+
+  Type resultElementType = cast<RankedTensorType>(resultType).getElementType();
+
+  PoolSizeCalculator<avgPoolDims> poolSizeCalculator(self, sumPool, rewriter,
+                                                     loc);
+  Value avgPool =
+      rewriter
+          .create<linalg::GenericOp>(
+              loc, outputTensor.getType(), sumPool, outputTensor,
+              /*indexingMaps=*/indexingMapsAvg,
+              /*iteratorTypes=*/iteratorTypesAvg,
+              [&](OpBuilder &b, Location loc, ValueRange args) {
+                auto poolSize = poolSizeCalculator.getPoolSize(
+                    b, kernelSizeIntValues, strideInts, paddingInts);
+                // AtenAvgPool2/3dOp has an optional divisor_override
+                // attribute while AtenAvgPool1dOp does not.
+                if constexpr (avgPoolDims > 1) {
```

Comment:
FYI this led to some downstream build failures on Windows: https://github.com/llvm/torch-mlir/issues/4085